service/mapper/transitions.go

```diff
@@ -269,9 +271,124 @@ func (t *Transitions) ResumeIndexing(s *State) error {
+	if errors.Is(err, archive.ErrUnavailable) {
+		time.Sleep(t.cfg.WaitInterval)
+		log.Debug().Msg("waiting for next trie update")
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("unable to retrieve trie updates for block height %x", s.height)
+	}
+	s.updates = updates
+	log.Info().Int("updates", len(s.updates)).Msg("collected trie updates (registers) to be mapped")
+
+	// Now that we have collected TrieUpdates for the current block,
+	// we can create a path -> payload mapping that we will use to index to the local storage
+	s.status = StatusCollect
+	return nil
+}
+
+// CollectRegisters reads the payloads for the next block to be indexed from the state's forest, unless payload
+// indexing is disabled.
+func (t *Transitions) CollectRegisters(s *State) error {
+	if s.status != StatusCollect {
+		return fmt.Errorf("invalid status for collecting registers (%s)", s.status)
+	}
+
+	// If indexing payloads is disabled, we can bypass collection and indexing
+	// of payloads and just go straight to indexing chain data
+	if t.cfg.SkipRegisters {
+		err := t.write.LatestRegisterHeight(s.height)
+		if err != nil {
+			return fmt.Errorf("could not index latest register height: %w", err)
+		}
+		s.status = StatusIndex
+		return nil
+	}
+	// collect paths/payload combinations
+	for _, update := range s.updates {
+		// guard for bootstrap case where t.updates.AllUpdates() returns nil as the queue is empty
+		if update != nil {
+			for i, path := range update.Paths {
+				s.registers[path] = update.Payloads[i]
+			}
+		}
+	}
+
+	// At this point, we have collected all the payloads, so we go to the next
+	// step, where we will index them.
+	s.status = StatusMap
+	return nil
+}
+
+// MapRegisters maps the collected registers to the current block.
+func (t *Transitions) MapRegisters(s *State) error {
+	if s.status != StatusMap {
+		return fmt.Errorf("invalid status for indexing registers (%s)", s.status)
+	}
+
+	log := t.log.With().Uint64("height", s.height).Logger()
+	// If there are no registers left to be indexed, we can go to the next step,
+	// which is indexing the chain data
+	if len(s.registers) == 0 {
+		err := t.write.LatestRegisterHeight(s.height)
+		if err != nil {
+			return fmt.Errorf("could not index latest register height: %w", err)
+		}
+		s.status = StatusIndex
+		return nil
+	}
+
+	// We will now collect and index 1000 registers at a time. This gives the
+	// FSM the chance to exit the loop between every 1000 payloads we index. It
+	// doesn't really matter for badger if they are in random order, so this
+	// way of iterating should be fine.
+	n := 1000
+	payloads := make([]*ledger.Payload, 0, n)
+	for path, payload := range s.registers {
+		payloads = append(payloads, payload)
+		delete(s.registers, path)
+		if len(payloads) >= n {
+			break
+		}
+	}
+
+	// Then we store the (maximum) 1000 paths and payloads.
+	err := t.write.Payloads(s.height, payloads)
+	if err != nil {
+		return fmt.Errorf("could not index registers: %w", err)
+	}
```

Comment:
nit: instead of reruning `MapRegisters()` for every 1000 payloads, maybe batch write payloads in the for loop until all registers are indexed.

With suggested approach:
- `payloads` slice is reused for all registers
- only need to update height and next status in one place afterwards.
- etc.

```suggestion Go        
	n := 1000
	if len(s.registers) < n {
	        n = len(s.registers)
	}

        payloads := make([]*ledger.Payload, 0, n)

        for path, payload := range s.registers {
	        payloads = append(payloads, payload)

		if len(payloads) >= n {
	                // Store max number of paths and payloads.
	                err := t.write.Payloads(s.height, payloads)
	                if err != nil {
		                return fmt.Errorf("could not index registers: %w", err)
	                }

	                // Reslice payloads for reuse.
	                payloads = payloads[:0]
                }
        }

        // Store remaining paths and payloads.
        if len(payloads) > 0 {
	        err := t.write.Payloads(s.height, payloads)
	        if err != nil {
		        return fmt.Errorf("could not index registers: %w", err)
	        }
        }
```